File: scala/leet_code/205.scala
/** This module checks if two given strings are isomorphic.
  * Two strings are isomorphic if the characters in one string can be replaced to get the other string.
  * All occurrences of a character must be replaced with another character while preserving the order of characters.
  * No two characters may map to the same character, but a character may map to itself.
  * The function `isIsomorphic` is the main function in this module.
  */
object Solution {

  /** Function to check if two given strings are isomorphic.
    *
    * @param s a String representing the first input string to be checked.
    * @param t a String representing the second input string to be checked.
    * @return a Boolean value indicating whether the two input strings are isomorphic. Returns true if the strings are isomorphic, false otherwise.
    */
  def isIsomorphic(s: String, t: String): Boolean = {
    // Initialize a mutable map to keep track of the character mappings
    val map = scala.collection.mutable.Map[Char, Char]()
    // Initialize a mutable set to keep track of the characters that have already been mapped
    val set = scala.collection.mutable.Set[Char]()

    // Use the `exists` method to check if there exists an index in the strings where the isomorphism condition is violated
    !(0 until s.length).exists { i =>
      // If the character in `s` at the current index has already been mapped
      if (map.contains(s(i))) {
        // Check if the mapped character is not equal to the character in `t` at the current index
        map(s(i)) != t(i)
      } else {
        // If the character in `t` at the current index has already been mapped
        if (set.contains(t(i))) {
          // The isomorphism condition is violated
          true
        } else {
          // Add the character mapping to the map and the character in `t` at the current index to the set
          map(s(i)) = t(i)
          set += t(i)
          // The isomorphism condition is not violated
          false
        }
      }
    }
  }
}

File: javascript/leet_code/205.mjs
/**
 * Function to determine if two strings are isomorphic.
 * Two strings are isomorphic if the characters in one string can be replaced to get the other string.
 *
 * @param {string} s - The first string
 * @param {string} t - The second string
 * @returns {boolean} - True if the strings are isomorphic, false otherwise
 */
const isIsomorphic = function (s, t) {
    // Create maps to store the mapping from characters in s to characters in t and vice versa
    const sMap = new Map();
    const tMap = new Map();

    // Iterate over the characters in the strings
    for (let i = 0; i < s.length; i++) {
        // If the current character in s is already mapped to a character in t and it is not the same as the current
        // character in t
        if (sMap.has(s[i]) && sMap.get(s[i]) !== t[i]) {
            return false;
        }
        // If the current character in t is already mapped to a character in s and it is not the same as the current
        // character in s
        if (tMap.has(t[i]) && tMap.get(t[i]) !== s[i]) {
            return false;
        }
        // Map the current characters in s and t to each other
        sMap.set(s[i], t[i]);
        tMap.set(t[i], s[i]);
    }

    // If all characters in s and t can be mapped to each other, the strings are isomorphic
    return true;
};
